airflow/cli/commands/dag_command.py

```diff
@@ -460,19 +478,27 @@ def dag_test(args, session=None):
     execution_date = args.execution_date or timezone.utcnow()
     dag = get_dag(subdir=args.subdir, dag_id=args.dag_id)
     dag.clear(start_date=execution_date, end_date=execution_date, dag_run_state=False)
-    try:
-        dag.run(
-            executor=DebugExecutor(),
-            start_date=execution_date,
-            end_date=execution_date,
-            conf=run_conf,
-            # Always run the DAG at least once even if no logical runs are
-            # available. This does not make a lot of sense, but Airflow has
-            # been doing this prior to 2.2 so we keep compatibility.
-            run_at_least_once=True,
-        )
-    except BackfillUnfinished as e:
-        print(str(e))
+
+    dr: DagRun = _get_or_create_dagrun(
+        dag=dag,
+        start_date=args.execution_date,
+        execution_date=args.execution_date,
+        run_id=DagRun.generate_run_id(DagRunType.MANUAL, execution_date),
+        session=session,
+        conf=run_conf,
+    )
+
+    tasks = dag.task_dict
+    log.info("starting dagrun")
+    # Instead of starting a scheduler, we run the minimal loop possible to check
+    # for task readiness and dependency management. This is notably faster
+    # than creating a BackfillJob and allows us to surface logs to the user
+    while dr.state == State.RUNNING:
+        schedulable_tis, _ = dr.update_state(session=session)
+        for ti in schedulable_tis:
+            add_logger_if_needed(ti)
+            ti.task = tasks[ti.task_id]
+            _run_task(ti)
```

Comment:
General rule: don't use `@provide_session` anywhere but the "outside" of an API so

```suggestion
            _run_task(ti, session=session)
```